library/VersionInfoFactory.cpp

```diff
@@ -51,31 +51,26 @@ VersionInfoFactory::~VersionInfoFactory()
 
 void VersionInfoFactory::clear()
 {
-    // for each stored version, delete
-    for(size_t i = 0; i < versions.size();i++)
-    {
-        delete versions[i];
-    }
     versions.clear();
     error = false;
 }
 
-VersionInfo * VersionInfoFactory::getVersionInfoByMD5(string hash)
+const VersionInfo * VersionInfoFactory::getVersionInfoByMD5(string hash)
 {
     for(size_t i = 0; i < versions.size();i++)
     {
         if(versions[i]->hasMD5(hash))
-            return versions[i];
+            return versions[i].get();
```

Comment:
If you use shared_ptr you could return shared_ptr or shared_ptr reference here. Then callers need to just copy the shared_ptr instead of VersionInfo structure.

Alternative would be to use unique_ptr in versions vector. You can insert unique_ptr to container with versions.emplace_back(std::move(version));

I haven't studied the code enough to recommend either approach. But if vinfo is modified by caller then I suspect the later is better solution while in other cases copy of shared_ptr is cheaper than copy of VersionInfo.

Reply:
I do want to modify this further to remove some more of the pointers and new/delete calls but I wanted to do it one thing at a time.